src/main/java/org/truffleruby/core/encoding/EncodingNodes.java

```diff
@@ -78,7 +78,19 @@ public static NegotiateCompatibleRopeEncodingNode create() {
             return NegotiateCompatibleRopeEncodingNodeGen.create();
         }
 
-        @Specialization(guards = "firstEncoding == secondEncoding")
+        @Specialization(guards = {
+                "firstEncoding == secondEncoding",
+                "firstEncoding == cachedEncoding",
```

Comment:
I think it's slightly better to compare both to `cachedEncoding`, compare both to a constant.

Reply:
If it's a performance matter, I don't mind doing it. Otherwise, I think as a guard it's clearer that this specialization is run when the two encodings are the same. Comparing them both to a cached value is a roundabout way of asserting that.

Reply:
I'm not entirely sure re performance, it's probably close but I would think a little bit better.
I like comparing both to a constant because then it's more symmetrical, and in the compiler graph we'd have something like `a == UTF-8 && b == UTF-8` instead of `a == b && a == UTF-8` which then from that POV looks like a roundabout way.